### ISSUE TITLE
Embedding in an open-source hypervisor impossible with current license?

### ISSUE THREAD

**Issue by stormi** [microsoft/secureboot_objects#138]
I am the lead maintainer and release manager for XCP-ng, an open-source virtualization server currently in its version 8.3.

Among our users’ needs is Secure Boot for guests. To enable it, the VMs require the appropriate certificate databases in their virtual NVRAM: `PK` (provided by us), `KEK`, `db`, and `dbx`. If these are absent, the VM loads them from the hypervisor’s disk. But first, they need to be present on the hypervisor's disk.

Currently, based on the license for the Secure Boot objects as described in the README file, our understanding is that we cannot freely distribute these files with our installer and RPM packages. This is because the restrictions in your license terms impose additional conditions that are incompatible with open-source licenses.  

For example:  

> you may distribute the Secure Boot Objects to end users **solely as part of the distribution of an operating system software product**, or as part of the distribution of updates to an operating system software product; and you may distribute the Secure Boot Objects to end users or through your distribution channels **solely as embodied in a firmware product or hardware product** that embodies nontrivial additional functionality  

We do not distribute an operating system in the traditional sense. Instead, we provide a virtualization platform. Each VM has its own operating system.  

Additionally, there is no firmware product or hardware product involved in our case, as our product is entirely *software*, functioning to some extent as hardware and firmware, like any virtualization system. For instance, we use EDK2 to provide virtual firmware to UEFI VMs.

> copying or reproduction of the Secure Boot Objects to any other server or location for further reproduction or redistribution on a standalone basis is expressly prohibited  

Although we would not distribute the objects on a standalone basis, these restrictions (and those highlighted above) conflict with the principles of Free Software and its licenses. We cannot impose additional constraints on our users, such as "you can use our product, but you cannot redistribute parts of it because Microsoft’s terms forbid it."  

This significantly hinders our ability to deliver a seamless experience to our users, who must individually download the Secure Boot objects to add them to their XCP-ng servers.  

**I hope that the license terms described in the README file have been superseded by the BSD license included in License.txt.** Please confirm whether this is the case. A fully free license for these objects would be excellent news.

**Comment by Flickdm** [comment]
@stormi I'm looking into this and will follow back up as soon as possible!

**Comment by stormi** [comment]
Dear bot, this issue is still valid, you can remove the stale label.

**Comment by Flickdm** [comment]
Sorry - removed the stale issue. As an update - still discussing with lawyers (and awaiting a reply).

(All of the following is "I am not a lawyer")

> solely as part of the distribution of an <s>operating system</s> software product, or
> as part of the distribution of updates to an <s>operating system</s> software product.

I was able to get them to remove the following from the license which would cover specifically the contents in [PostSignedObjects ](https://github.com/microsoft/secureboot_objects/tree/main/PostSignedObjects) and cover the releases.

It is my understanding that the BSD license only covers the code (python) and the [CSV -like objects](https://github.com/microsoft/secureboot_objects/blob/main/PreSignedObjects/DBX/Hashes/dbx_info_uefi_org_7_18_23.csv) and the [config file](https://github.com/microsoft/secureboot_objects/blob/main/FirmwareDefaults.toml). Thus it is possible for a downstream to use the these to generate the defaults under the BSD license. 

It's just that we're trying to limit exposure for the compiled binaries. This is the part I'm still discussing to see if there is something else we can do here.

**Comment by stormi** [comment]
Thanks for following-up.

> (If you think this would help you I can make a PR to make this change immediately)

This would likely be better, but this would still be too restrictive for embedding in an open source hypervisor distribution like ours.

**Comment by Flickdm** [comment]
Yeah that's what I imagined. In my understanding the limited liability licensing is meant to cover the compiled binaries (both unsigned and signed). However, I don't believe that licensing applies to the CSV or python script that can generate the unsigned binaries. While admittedly silly would it be of any use to you run the script and generate your own unsigned binary and thus be covered under your own licensing?

**Comment by pbatard** [comment]
I'm going to piggyback on this issue as I am the developer a [Mosby](https://github.com/pbatard/Mosby), a UEFI Shell application that helps users update/replace their Secure Boot database binaries (including support for the 2011 certs revocation/expiration) as well as generate/install their own Secure Boot credentials.

As part of this, I obviously want to embed the various DBXs within the software application itself, which the current license allegedly makes a grey area (though I have to say that, considering that nobody here is trying to abuse/impersonate/profit from what has to be, by design, a noncommercial public list of hashes, I don't see how the current legalese would stand in a court of law tasked to examine the justification of restricting redistribution of unmodified public DBX files to OS manufacturers only).

Therefore, whereas I do not have the same requirements as OP, I would greatly appreciate if the proposed change of:

> solely as part of the distribution of an ~~operating system~~ software product, or
> as part of the distribution of updates to an ~~operating system~~ software product.

could proceed forward, even as some more work might still be needed to resolve OP's problem. As far as my software product is concerned, I believe that this is all that's needed to make sure that any uncertainty, regarding its legal compliance with regards to its ability to install DBXs, is lifted.

Thank you.

**Comment by pbatard** [comment]
@Flickdm, thank you for applying the license changes!

**Comment by stormi** [comment]
>  While admittedly silly would it be of any use to you run the script and generate your own unsigned binary and thus be covered under your own licensing?

Our users can already do that if they want, but for the rest of our users, would it still allow to boot original Windows VMs with Secure Boot (and to let Windows Update update the dbx)?

**Comment by Flickdm** [comment]
@stormi As long as your L"KEK" variable in firmware has the [Microsoft KEK Certificates](https://github.com/microsoft/secureboot_objects/tree/main/PreSignedObjects/KEK/Certificates) in it then Windows (or Linux) should be able to service it regardless of what was in there to begin with. 

The unsigned binaries are just a way to update the defaults in firmware more consistently and accurately however those only apply if someone is enabling secure boot for the first time or have reset secure boot at some point. The OS is really the method by which the platform should keep them up to date.

**Comment by dinhngtu** [comment]
Could you please confirm the license/distribution terms of the certificates in the [PreSignedObjects](https://github.com/microsoft/secureboot_objects/tree/main/PreSignedObjects) directory?

**Comment by stormi** [comment]
Thank you for your diligence, annoying github action, but could we let humans decide when a issue is stale?

**Comment by pbatard** [comment]
@stormi,

I think you want to modify or remove your https://github.com/microsoft/secureboot_objects/blob/main/.github/workflows/stale.yml, that references https://github.com/microsoft/mu_devops/blob/main/.github/workflows/Stale.yml, and that seems to be controlling this behaviour.

I don't believe that GitHub automatically closes stale issues on its own, especially as I have always had to use custom workflows like this on repos where I want what stale issues to be closed.

**Comment by makubacki** [comment]
The stale issue is not used to discard issues but to bring attention that it exists. A human is supposed to be notified (as happened here) and remove the stale label if it is not stale. Hopefully taking that as an opportunity to make forward progress on the issue.

**Comment by makubacki** [comment]
@Flickdm, are you still active to respond to the last comment?

> Could you please confirm the license/distribution terms of the certificates/dbx JSON files in the [PreSignedObjects](https://github.com/microsoft/secureboot_objects/tree/main/PreSignedObjects) directory?

**Comment by Flickdm** [comment]
@dinhngtu I can confirm that the code artifacts being provided in the repository are licensed under BSD. You are free to fork this repository and make changes as you see fit.

**Comment by Flickdm** [comment]
I think this issue has diverged a bit. @stormi are you satisfied with the conclusion of this issue?
